hummingbot/connector/exchange/binance/binance_api_order_book_data_source.py

```diff
@@ -152,64 +145,91 @@ async def get_new_order_book(self, trading_pair: str) -> OrderBook:
+                                  f"Error: {e}")
+            raise
+
+    async def _iter_messages(self,
+                             ws: aiohttp.ClientWebSocketResponse) -> AsyncIterable[Any]:
         try:
             while True:
-                try:
-                    msg: str = await asyncio.wait_for(ws.recv(), timeout=self.MESSAGE_TIMEOUT)
-                    yield msg
-                except asyncio.TimeoutError:
-                    pong_waiter = await ws.ping()
-                    await asyncio.wait_for(pong_waiter, timeout=self.PING_TIMEOUT)
-        except asyncio.TimeoutError:
-            self.logger().warning("WebSocket ping timed out. Going to reconnect...")
-            return
-        except ConnectionClosed:
-            return
+                yield await ws.receive_json()
+        except asyncio.CancelledError:
+            raise
+        except Exception as e:
+            self.logger().network(f"Unexpected error occured when parsing websocket payload. "
+                                  f"Error: {e}")
+            raise
         finally:
             await ws.close()
 
     async def listen_for_trades(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+        ws = None
         while True:
             try:
-                ws_path: str = "/".join([f"{binance_utils.convert_to_exchange_trading_pair(trading_pair).lower()}@trade"
-                                         for trading_pair in self._trading_pairs])
-                url = CONSTANTS.WSS_URL.format(self._domain)
-                stream_url: str = f"{url}/{ws_path}"
-
-                ws = await websockets.connect(stream_url)
-                async for raw_msg in self._inner_messages(ws):
-                    msg = ujson.loads(raw_msg)
-                    trade_msg: OrderBookMessage = BinanceOrderBook.trade_message_from_exchange(msg)
+                ws = await self._create_websocket_connection()
+                payload = {
+                    "method": "SUBSCRIBE",
+                    "params":
+                    [
+                        f"{binance_utils.convert_to_exchange_trading_pair(trading_pair).lower()}@trade"
+                        for trading_pair in self._trading_pairs
+                    ],
+                    "id": self.TRADE_STREAM_ID
+                }
+                await ws.send_json(payload)
+
+                async for json_msg in self._iter_messages(ws):
+                    if "result" in json_msg:
+                        continue
+                    trade_msg: OrderBookMessage = BinanceOrderBook.trade_message_from_exchange(json_msg)
                     output.put_nowait(trade_msg)
             except asyncio.CancelledError:
                 raise
             except Exception:
                 self.logger().error("Unexpected error with WebSocket connection. Retrying after 30 seconds...",
                                     exc_info=True)
+                ws and await ws.close()
```

Comment:
This line should be inside a `finally` statement, not only in the `except Exception`